readium/lcp/src/main/java/org/readium/r2/lcp/LcpDecryptor.kt

```diff
@@ -17,30 +17,40 @@ import org.readium.r2.shared.extensions.coerceFirstNonNegative
 import org.readium.r2.shared.extensions.inflate
 import org.readium.r2.shared.extensions.requireLengthFitInt
 import org.readium.r2.shared.fetcher.*
-import org.readium.r2.shared.publication.Link
-import org.readium.r2.shared.publication.encryption.encryption
+import org.readium.r2.shared.publication.encryption.Encryption
+import org.readium.r2.shared.resource.FailureResource
 import org.readium.r2.shared.resource.Resource
 import org.readium.r2.shared.resource.ResourceTry
+import org.readium.r2.shared.resource.TransformingResource
+import org.readium.r2.shared.resource.flatMap
+import org.readium.r2.shared.resource.flatMapCatching
+import org.readium.r2.shared.resource.mapCatching
+import org.readium.r2.shared.util.Url
 
 /**
  * Decrypts a resource protected with LCP.
  */
-internal class LcpDecryptor(val license: LcpLicense?) {
-
-    fun transform(resource: Fetcher.Resource): Fetcher.Resource = LazyResource {
-        // Checks if the resource is encrypted and whether the encryption schemes of the resource
-        // and the DRM license are the same.
-        val link = resource.link()
-        val encryption = link.properties.encryption
-        if (encryption == null || encryption.scheme != "http://readium.org/2014/01/lcp")
-            return@LazyResource resource
-
-        when {
-            license == null -> FailureResource(link, Resource.Exception.Forbidden())
-            link.isDeflated || !link.isCbcEncrypted -> FullLcpResource(resource, license)
-            else -> CbcLcpResource(resource, license)
+internal class LcpDecryptor(
```

Comment:
`LcpDecryptor` don't require the encryption metadata from the input `Link` anymore. Instead, we have an explicit `retrieveEncryption` dependency which will return the encryption metadata associated with a source URL.